source/dem/particle_particle_contact_force.cc

```diff
@@ -17,25 +17,984 @@
+
+              this->model_parameter_beta[i].insert(
+                {j,
+                 restitution_coefficient_particle_log /
+                   sqrt(restitution_coefficient_particle_log *
+                          restitution_coefficient_particle_log +
+                        9.8696)});
+            }
+        }
+    }
+
+  if (dem_parameters.model_parameters.rolling_resistance_method ==
+      Parameters::Lagrangian::ModelParameters::RollingResistanceMethod::
+        no_resistance)
+    this->rolling_resistance_model =
+      RollingResistanceTorqueModel::no_rolling_resistance;
+  else if (dem_parameters.model_parameters.rolling_resistance_method ==
+           Parameters::Lagrangian::ModelParameters::RollingResistanceMethod::
+             constant_resistance)
+    this->rolling_resistance_model =
+      RollingResistanceTorqueModel::constant_rolling_resistance;
+  else if (dem_parameters.model_parameters.rolling_resistance_method ==
+           Parameters::Lagrangian::ModelParameters::RollingResistanceMethod::
+             viscous_resistance)
+    this->rolling_resistance_model =
+      RollingResistanceTorqueModel::viscous_rolling_resistance;
 }
 
-template class ParticleParticleContactForce<2>;
-template class ParticleParticleContactForce<3>;
+template <
+  int                                                       dim,
+  Parameters::Lagrangian::ParticleParticleContactForceModel contact_model>
+void
+ParticleParticleContactForce<dim, contact_model>::
+  calculate_particle_particle_contact_force(
+    typename DEM::dem_data_structures<dim>::adjacent_particle_pairs
+      &local_adjacent_particles,
+    typename DEM::dem_data_structures<dim>::adjacent_particle_pairs
+      &                        ghost_adjacent_particles,
+    const double               dt,
+    std::vector<Tensor<1, 3>> &torque,
+    std::vector<Tensor<1, 3>> &force)
+{
+  // Contact forces calculations of local-local and local-ghost particle
+  // pairs are performed in separate loops
+
+  // Looping over local_adjacent_particles values with iterator
+  // adjacent_particles_list
+  for (auto &&adjacent_particles_list :
+       local_adjacent_particles | boost::adaptors::map_values)
+    {
+      if (!adjacent_particles_list.empty())
+        {
+          // Gather information about particle 1 and set it up.
+          auto first_contact_info = adjacent_particles_list.begin();
+          auto particle_one       = first_contact_info->second.particle_one;
+          auto particle_one_properties = particle_one->get_properties();
+
+          types::particle_index particle_one_id =
+            particle_one->get_local_index();
+          Tensor<1, 3> &particle_one_torque = torque[particle_one_id];
+          Tensor<1, 3> &particle_one_force  = force[particle_one_id];
+
+          // Fix particle one location for 2d and 3d
+          Point<3> particle_one_location = [&] {
+            if constexpr (dim == 3)
```

Comment:
As in other parts of the code, I would explicitly say that dim should be either == 3 or == 2 instead of using "else", for defensive purposes.